### PR TITLE
deps: update libcurl to 7.76.0.

### PR DESCRIPTION
**Pull request by htuch**
Also fix CPE to point at libcurl rather than curl.

Signed-off-by: Harvey Tuch <htuch@google.com>

### CODE REVIEW

**Comment by htuch** [comment]
@wrowe can you take a look at this from patch perspective?

File: bazel/repository_locations.bzl [hunk 1]
```diff
@@ -761,8 +761,8 @@ REPOSITORY_LOCATIONS_SPEC = dict(
             "envoy.grpc_credentials.aws_iam",
             "envoy.tracers.opencensus",
         ],
-        release_date = "2020-12-09",
-        cpe = "cpe:2.3:a:haxx:curl:*",
+        release_date = "2021-03-31",
```
Comment (wrowe) [inline bazel/repository_locations.bzl:764]: It appears you need the 30th here.

File: bazel/foreign_cc/curl.patch [hunk 2]
```diff
@@ -67,14 +67,3 @@ index 6a1a6fe8e..777ee122f 100644
...
- 
- # Check for idn
--check_library_exists_concat("idn2" idn2_lookup_ul HAVE_LIBIDN2)
-+option(USE_LIBIDN2 "Use libidn2 for IDN support" ON)
-+set(HAVE_LIBIDN2 OFF)
-+if(USE_LIBIDN2)
-+  check_library_exists_concat("idn2" idn2_lookup_ul HAVE_LIBIDN2)
-+endif()
- 
- # Check for symbol dlopen (same as HAVE_LIBDL)
```
Comment (wrowe) [inline bazel/foreign_cc/curl.patch:80]: This looks correct, now incorporated in the release with https://github.com/curl/curl/commit/83f1ca69290f9d6198fe4013488408f94736c5dc

The corresponding USE_WIN32_IDN change defaults to off, which is our desired omission.

**Review by wrowe — APPROVED** [review]
LGTM, need to see CI results.

File: bazel/repository_locations.bzl (hunk 1 above)
Comment (wrowe) [reply bazel/repository_locations.bzl:764]: It's odd, that this was 00:08 GMT on the 31st - it seems really unwise that bazel build is using local time for package validation.

Comment (moderation) [reply bazel/repository_locations.bzl:764]: I'm very close to either writing or finding a browser extension that will convert Github times to UTC, but for now I inspect the release date and get the `datetime` value from the source

image

**Review by moderation — APPROVED** [review]

File: bazel/repository_locations.bzl (hunk 1 above)
Comment (wrowe) [reply bazel/repository_locations.bzl:764]: The date value expected by bazel-envoy build is 2021-03-30, if you inspect the failure log. Which only makes sense if CI times are PDT or later.

**Comment by wrowe** [comment]
@htuch it's on you to bump the date back one day

```
com_github_curl has a GitHub release date 2021-03-30
An error occurred while processing ./bazel/repository_locations.bzl, please verify the correctness of the metadata: Mismatch with metadata release date of 2021-03-31
##[error]Bash exited with code '1'.
```

**Comment by moderation** [comment]
The way the release date is retrieved today at https://github.com/envoyproxy/envoy/blob/main/tools/dependency/release_dates.py#L58-L69 is incorrect. It says the release date is the that of the latest commit. However releases can be cut at times later than the last commit and this release date is what is shown in the UI. For this curl release the actual UTC release date is `2021-03-31`

Here is my code that makes a different API call to get the actual UTC release date:
```python
# Extract release date from GitHub API.
def get_release_date(repo, metadata_version, github_release):
    if github_release.tagged:

        try:
            latest = repo.get_latest_release()
        except github.GithubException as e:
            # print(Fore.RED + 'GitHub API status get_latest_release(): {}'.format(str(e)) + Style.RESET_ALL)
            latest = ''
            pass

        if latest and (github_release.version <= latest.tag_name):
            # print("latest.tag_name: ", latest.tag_name)
            # print("github_release.version: ", github_release.version)
            release = repo.get_release(github_release.version)
            # print("release.published_at", release.published_at)
            return release.published_at
        else:
            tags = repo.get_tags()
            current_metadata_tag_commit_date = ''
            for tag in tags.reversed:
                # print("tag.name: ",version.parse(tag.name))
                # print("github_release.version",github_release.version)
                if tag.name == github_release.version:
                    # return tag.commit.commit.committer.date
                    current_metadata_tag_commit_date = tag.commit.commit.committer.date
                if not (version.parse(tag.name).is_prerelease) and version.parse(
                    tag.name
                ) > version.parse(github_release.version):
                    print(
                        Fore.YELLOW
                        + f'*WARNING* {repo.name} has a newer release than {github_release.version}@<{current_metadata_tag_commit_date}>: '
                        f'{tag.name}@<{tag.commit.commit.committer.date}>'
                        + Style.RESET_ALL
                    )
            return current_metadata_tag_commit_date
        return None
```
I need to fixup my commit checking and remove the debug and will put in a PR for this (and fix the many incorrect non-UTC dates in `bazel/repository_locations.bzl`

**Review by snowp — APPROVED** [review]

**Comment by wrowe** [comment]
> I need to fixup my commit checking and remove the debug and will put in a PR for this (and fix the many incorrect non-UTC dates in `bazel/repository_locations.bzl`

I'd noticed that some time ago, but we frequently include-by-tag, so that has to be retained (in addition to, if not in place of.)

Thanks for looking more closely at this, getting this to UTC will be a big help.

**Pull request merged**